In [1]:
import pandas as pd
import numpy as np


cell_phones = pd.read_csv('cell_phones.csv')
population = pd.read_csv('population1.csv')
geo_code = pd.read_csv('geo-code.csv')


# population.head()
# cell_phones.head()
geo_code.head()

country g77_and_oecd_countries income_3groups        income_groups  \
0    abkh                 others            NaN                  NaN   
1     abw                 others    high_income          high_income   
2     afg                    g77     low_income           low_income   
3     ago                    g77  middle_income  lower_middle_income   
4     aia                 others            NaN                  NaN   

   is--country iso3166_1_alpha2 iso3166_1_alpha3  iso3166_1_numeric iso3166_2  \
0         True              NaN              NaN                NaN       NaN   
1         True               AW              ABW              533.0       NaN   
2         True               AF              AFG                4.0       NaN   
3         True               AO              AGO               24.0       NaN   
4         True               AI              AIA              660.0       NaN   

   landlocked  ...  longitude  main_religion_2008         name  \
0         NaN  ...        NaN                 NaN     Abkhazia   
1   coastline  ...  -69.96667           christian        Aruba   
2  landlocked  ...   66.00000              muslim  Afghanistan   
3   coastline  ...   18.50000           christian       Angola   
4   coastline  ...  -63.05000           christian     Anguilla   

               un_sdg_ldc                       un_sdg_region un_state  \
0                     NaN                                 NaN    False   
1  un_not_least_developed  un_latin_america_and_the_caribbean    False   
2      un_least_developed        un_central_and_southern_asia     True   
3      un_least_developed               un_sub_saharan_africa     True   
4  un_not_least_developed  un_latin_america_and_the_caribbean    False   

   unicef_region unicode_region_subtag world_4region        world_6region  
0            NaN                   NaN        europe  europe_central_asia  
1            NaN                    AW      americas              america  
2             sa                    AF          asia           south_asia  
3            ssa                    AO        africa   sub_saharan_africa  
4            NaN                    AI      americas              america  

[5 rows x 21 columns]

In [2]:
cell_phones.head()

Country  year  cell_phones_total
0  Afghanistan  1960                0.0
1  Afghanistan  1961                0.0
2  Afghanistan  1962                0.0
3  Afghanistan  1963                0.0
4  Afghanistan  1964                0.0

In [3]:
geo_code.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 273 entries, 0 to 272
Data columns (total 21 columns):
country                   273 non-null object
g77_and_oecd_countries    259 non-null object
income_3groups            218 non-null object
income_groups             218 non-null object
is--country               273 non-null bool
iso3166_1_alpha2          246 non-null object
iso3166_1_alpha3          247 non-null object
iso3166_1_numeric         247 non-null float64
iso3166_2                 1 non-null object
landlocked                255 non-null object
latitude                  241 non-null float64
longitude                 241 non-null float64
main_religion_2008        216 non-null object
name                      273 non-null object
un_sdg_ldc                249 non-null object
un_sdg_region             248 non-null object
un_state                  273 non-null bool
unicef_region             195 non-null object
unicode_region_subtag     248 non-null object
world_4region           

In [4]:
cell_phones.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11640 entries, 0 to 11639
Data columns (total 3 columns):
Country              11640 non-null object
year                 11640 non-null int64
cell_phones_total    11640 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 272.9+ KB


In [5]:
population.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13651 entries, 0 to 13650
Data columns (total 3 columns):
Country       13651 non-null object
year          13651 non-null int64
population    13651 non-null int64
dtypes: int64(2), object(1)
memory usage: 320.1+ KB


CELL PHONE IS MATCHED WITH POPULATION SUCH THAT WE HAVE 11640 ROWS AND 4 COLUMNS

In [6]:
joined_df = cell_phones.merge(population, how='left', on=('Country', 'year'))


joined_df.head()

Country  year  cell_phones_total  population
0  Afghanistan  1960                0.0   9000000.0
1  Afghanistan  1961                0.0   9170000.0
2  Afghanistan  1962                0.0   9350000.0
3  Afghanistan  1963                0.0   9540000.0
4  Afghanistan  1964                0.0   9740000.0

In [7]:
joined_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11640 entries, 0 to 11639
Data columns (total 4 columns):
Country              11640 non-null object
year                 11640 non-null int64
cell_phones_total    11640 non-null float64
population           2700 non-null float64
dtypes: float64(2), int64(1), object(1)
memory usage: 454.7+ KB


RENAMING COLUMNS IN THE GEO_CODE DATAFRAME

In [8]:
geo_code.rename(columns={'country':'geo', 'name':'Country'}, inplace=True)

geo_code.head()

geo g77_and_oecd_countries income_3groups        income_groups  \
0  abkh                 others            NaN                  NaN   
1   abw                 others    high_income          high_income   
2   afg                    g77     low_income           low_income   
3   ago                    g77  middle_income  lower_middle_income   
4   aia                 others            NaN                  NaN   

   is--country iso3166_1_alpha2 iso3166_1_alpha3  iso3166_1_numeric iso3166_2  \
0         True              NaN              NaN                NaN       NaN   
1         True               AW              ABW              533.0       NaN   
2         True               AF              AFG                4.0       NaN   
3         True               AO              AGO               24.0       NaN   
4         True               AI              AIA              660.0       NaN   

   landlocked  ...  longitude  main_religion_2008      Country  \
0         NaN  ...        NaN                 NaN     Abkhazia   
1   coastline  ...  -69.96667           christian        Aruba   
2  landlocked  ...   66.00000              muslim  Afghanistan   
3   coastline  ...   18.50000           christian       Angola   
4   coastline  ...  -63.05000           christian     Anguilla   

               un_sdg_ldc                       un_sdg_region un_state  \
0                     NaN                                 NaN    False   
1  un_not_least_developed  un_latin_america_and_the_caribbean    False   
2      un_least_developed        un_central_and_southern_asia     True   
3      un_least_developed               un_sub_saharan_africa     True   
4  un_not_least_developed  un_latin_america_and_the_caribbean    False   

   unicef_region unicode_region_subtag world_4region        world_6region  
0            NaN                   NaN        europe  europe_central_asia  
1            NaN                    AW      americas              america  
2             sa                    AF          asia           south_asia  
3            ssa                    AO        africa   sub_saharan_africa  
4            NaN                    AI      americas              america  

[5 rows x 21 columns]

In [9]:
# Creating a new df that comprises the geo and the Country columns from the geo_code dataframe
new_geo = geo_code[['geo', 'Country']]

new_geo.head()

geo      Country
0  abkh     Abkhazia
1   abw        Aruba
2   afg  Afghanistan
3   ago       Angola
4   aia     Anguilla

In [10]:
new_geo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 273 entries, 0 to 272
Data columns (total 2 columns):
geo        273 non-null object
Country    273 non-null object
dtypes: object(2)
memory usage: 4.4+ KB


In [11]:
# merging the joined_df with new_geo
final_joined_df = joined_df.merge(new_geo, how='inner', on='Country')

final_joined_df.head()

Country  year  cell_phones_total  population  geo
0  Afghanistan  1960                0.0   9000000.0  afg
1  Afghanistan  1961                0.0   9170000.0  afg
2  Afghanistan  1962                0.0   9350000.0  afg
3  Afghanistan  1963                0.0   9540000.0  afg
4  Afghanistan  1964                0.0   9740000.0  afg

In [12]:
final_joined_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11640 entries, 0 to 11639
Data columns (total 5 columns):
Country              11640 non-null object
year                 11640 non-null int64
cell_phones_total    11640 non-null float64
population           2700 non-null float64
geo                  11640 non-null object
dtypes: float64(2), int64(1), object(2)
memory usage: 545.6+ KB


In [15]:
# Creating a new feature ---> cell_phones_per_person

final_joined_df['cell_phones_per_person'] = final_joined_df['cell_phones_total']/final_joined_df['population']

In [16]:
final_joined_df.head()

Country  year  cell_phones_total  population  geo  \
0  Afghanistan  1960                0.0   9000000.0  afg   
1  Afghanistan  1961                0.0   9170000.0  afg   
2  Afghanistan  1962                0.0   9350000.0  afg   
3  Afghanistan  1963                0.0   9540000.0  afg   
4  Afghanistan  1964                0.0   9740000.0  afg   

   cell_phones_per_person  
0                     0.0  
1                     0.0  
2                     0.0  
3                     0.0  
4                     0.0

In [18]:
# Converting the geo column to upper case

final_joined_df['geo'] = final_joined_df['geo'].str.upper()

final_joined_df.head()

Country  year  cell_phones_total  population  geo  \
0  Afghanistan  1960                0.0   9000000.0  AFG   
1  Afghanistan  1961                0.0   9170000.0  AFG   
2  Afghanistan  1962                0.0   9350000.0  AFG   
3  Afghanistan  1963                0.0   9540000.0  AFG   
4  Afghanistan  1964                0.0   9740000.0  AFG   

   cell_phones_per_person  
0                     0.0  
1                     0.0  
2                     0.0  
3                     0.0  
4                     0.0

In [19]:
final_joined_df.describe()

year  cell_phones_total    population  cell_phones_per_person
count  11640.000000       1.164000e+04  2.700000e+03             2700.000000
mean    1989.500000       7.875722e+06  3.852029e+07                0.227232
std       17.318846       5.572496e+07  1.680886e+08                0.415575
min     1960.000000       0.000000e+00  1.340000e+04                0.000000
25%     1974.750000       0.000000e+00  1.147500e+06                0.000000
50%     1989.500000       0.000000e+00  6.195000e+06                0.000000
75%     2004.250000       5.615000e+05  1.380000e+07                0.243566
max     2019.000000       1.730000e+09  1.430000e+09                2.090909

In [20]:
# Asserting that the dataset comprises of 194 countries

assert(len(final_joined_df['Country'].unique()) == 194)

In [21]:
# Saving countries in 2017 to a variable
countries_in_2017 = final_joined_df[['Country', 'cell_phones_total']].loc[final_joined_df['year']==2017]

countries_in_2017.head()

Country  cell_phones_total
57   Afghanistan         23900000.0
117      Albania          3630000.0
177      Algeria         45800000.0
237      Andorra            80300.0
297       Angola         13300000.0

In [22]:
# Getting the top 5 countries in 2017 in terms of cell_phones total
countries_in_2017 = countries_in_2017.sort_values(by=['cell_phones_total'], ascending=False)
top_5_countries_in_2017 = countries_in_2017.head()

top_5_countries_in_2017

Country  cell_phones_total
2157           China       1.470000e+09
4617           India       1.170000e+09
4677       Indonesia       4.350000e+08
11157  United States       4.000000e+08
8577          Russia       2.270000e+08